### PR TITLE
[build-tools] Merge tool 

### PR DESCRIPTION
**Pull request by sonalivdeshpande**
This PR includes the following changes to the main/next automation process:

1. Initially, automation was part of the GitHub workflow. It has been moved as part of our new build-cli 
2. This helps us to easier modified the automation process and keep GitHub workflow files as small as possible
3. I have eliminated use of Azure Queue and used git commands to get the last merged commit between main/next and find the list of unmerged commits. This was possible since we now merge commit instead of squash-merge.

`git merge-base main next`

`git rev-list COMMIT_ID..main`

4. The right merge conflict owner is assigned the automation PR.
5. Automation prs not getting assigned to external developers

### CODE REVIEW

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    };
+
+    public async run(): Promise<void> {
+        const flags = this.processedFlags;
+
+        const resolvedRoot = await getResolvedFluidRoot();
+        const gitRepo = new GitRepo(resolvedRoot);
+
+        if (await prExists(flags.auth, "Automation: Main-next integrate")) {
+            this.log(`Open pull request exists`);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:124]: maybe error or warning?

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts [hunk 2]
```diff
@@ -194,6 +194,40 @@ export class GitRepo {
         return date;
     }
 
+    /**
+     * @param source - Source branch name
+     * @param target: Target branch name
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:199]:
```suggestion
     * @param target - Target branch name
```

File: build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts [hunk 3]
```diff
@@ -194,6 +194,40 @@ export class GitRepo {
...
+        return await this.exec(`merge-base ${source} ${target}`, `merge base ${source} and ${target} branch`);
+    }
+
+    public async merge(commitId: string) {
+        return await this.exec(`merge ${commitId} --no-ff`, `merge a commit id`);
+    }
+
+    public async mergeAbort() {
+        return await this.exec(`merge --abort`, `abort the merge`);
+    }
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/bumpVersion/gitRepo.ts:212]: nit: maybe combine these into one method.

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            multiple: true,
+        }),
+        ...BaseCommand.flags,
+    };
+
+    public async run(): Promise<void> {
+        const flags = this.processedFlags;
+
+        const resolvedRoot = await getResolvedFluidRoot();
+        const gitRepo = new GitRepo(resolvedRoot);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:121]: I suggest using the context instead:

```suggestion
        const context = await this.getContext();
        const gitRepo = context.gitRepo;
        const resolvedRoot = gitRepo.resolvedRoot;
```

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+        const unmergedCommitList: string[] = [];
+        let str = "";
+
+        for (const id of listCommits) {
+            // check the length to be 40 since git commit id is 40 digits long
+            if (str.length === 40) {
+                unmergedCommitList.push(str);
+                str = "";
+                continue;
+            }
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:143]: I would output a warning if you see an unexpected string.

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+
+            str += id;
+        }
+
+        this.log(
+            `There are ${unmergedCommitList.length} unmerged commits between ${flags.source} and ${flags.target} branches`,
+        );
+
+        if (unmergedCommitList.length === 0) {
+            this.log(
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:153]: Consider using this.info instead of this.log.

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+
+        let prNumber;
+        if(i === 0) {
+            this.log(`Opening a pull request for a single commit id ${unmergedCommitList[0]} as it might have merge conflicts with the ${flags.target} branch.`);
+            await gitRepo.switchBranch(flags.source);
+            await gitRepo.createBranch(`${flags.source}-${flags.target}-${unmergedCommitList[0]}`);
+            await gitRepo.fetchBranch(flags.source, `${flags.source}-${flags.target}-${unmergedCommitList[0]}`);
+            await gitRepo.setUpstream(`${flags.source}-${flags.target}-${unmergedCommitList[0]}`);
+            await gitRepo.resetBranch(unmergedCommitList[0]);
+            prNumber = await createPR(flags.auth, `${flags.source}-${flags.target}-${unmergedCommitList[0]}`, flags.target, "sonalivdeshpande");
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:197]: should your username be here?

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+            // fetch name of owner associated to the pull request
+            const data = await prInfo(flags.auth, unmergedCommitList[0]);
+            this.log(`Fetch pull request info for commit id ${unmergedCommitList[0]}: ${data}`);
+            this.log(`Open pull request for commit id ${unmergedCommitList[0]}. Please resolve the merge conflicts.`);
+        } else {
+            prNumber = await createPR(flags.auth, branchName, flags.target, "sonalivdeshpande");
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:203]: also here

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+                break;
+            }
+
+            commit = i;
+            i++;
+        }
+
+        let prNumber;
+        if(i === 0) {
+            this.log(`Opening a pull request for a single commit id ${unmergedCommitList[0]} as it might have merge conflicts with the ${flags.target} branch.`);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:191]:
```suggestion
            this.log(`Opening a pull request for a single commit (${unmergedCommitList[0]}) because it might have merge conflicts with the ${flags.target} branch.`);
```

Could we say it does have conflicts rather than say "might"? Or do we not know yet?

File: build-tools/packages/build-cli/src/commands/merge/branch.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        let i = 0;
+        let commit = 0;
+        while(i !== commitSize) {
+            try {
+                // eslint-disable-next-line no-await-in-loop
+                await gitRepo.merge(unmergedCommitList[i]);
+            } catch (error: unknown) {
+                // eslint-disable-next-line no-await-in-loop
+                await gitRepo.mergeAbort();
+                this.log(`Merge abort for ${unmergedCommitList[i]}: `, error);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branch.ts:181]: warning or error?

File: build-tools/packages/build-cli/README.md [hunk 11]
```diff
@@ -340,6 +341,29 @@ DESCRIPTION
 
 _See code: [dist/commands/info.ts](https://github.com/microsoft/FluidFramework/blob/v0.4.5000/dist/commands/info.ts)_
 
+## `flub merge branch`
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/README.md:344]: I think `merge branches` is better than just "merge branch."

**Review by tylerbutler — COMMENTED** [review]
Looking good! Let some questions and feedback.

File: build-tools/packages/build-cli/src/commands/merge/branch.ts (hunk 7 above)
Comment (sonalivdeshpande) [reply build-tools/packages/build-cli/src/commands/merge/branch.ts]: Tested it locally and username is not required for any of the git apis. The only problem I'm seeing is `git rev-list` is not returning the right commit ids on my fork.

File: build-tools/packages/build-cli/src/api.ts [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+    return prInfo;
+}
+
+/**
+ *
+ * @param token - GitHub authentication token
+ * @returns Lists the user who have push access to this branch
+ */
+export async function getUserAccess(token: string): Promise<any> {
+    const octokit = new Octokit({ auth: token });
```
Comment (sonalivdeshpande) [inline build-tools/packages/build-cli/src/api.ts:65]: I was thinking of using this api for external developers. This api would return a list of users who have push access to the `main` branch.

File: .github/workflows/merge-branches.yml [hunk 13]
```diff
@@ -0,0 +1,15 @@
+name: Merge branches
+
+# on:
+#   schedule:
+#     - cron: "0 */2 * * 1-5" # Cron timezone is in UTC
+
```
Comment (tylerbutler) [inline .github/workflows/merge-branches.yml:6]: Go ahead and remove this.

File: build-tools/packages/build-cli/src/lib/dangerfile.ts [hunk 14]
```diff
@@ -2,7 +2,7 @@
  * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
  * Licensed under the MIT License.
  */
-import { Logger } from "@fluidframework/build-tools";
+import { defaultLogger, Logger } from "@fluidframework/build-tools";
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/dangerfile.ts:5]: Are these related changes?

**Review by tylerbutler — APPROVED** [review]

File: build-tools/packages/build-cli/src/lib/dangerfile.ts (hunk 14 above)
Comment (sonalivdeshpande) [reply build-tools/packages/build-cli/src/lib/dangerfile.ts:5]: These are not related to `merge-tool` but logger was not assigned

**Comment by ssimic2** [comment]
@sonalivdeshpande could you please describe the scope of this PR. It makes it easier to review it. Also, what are the use cases we concluded on that are driving MergeBranch implementation? I'm assuming you are trying to automate main/next here. I know we have discussed a couple of different approaches, I'm curious where we are at given the data we have so far.

**Comment by sonalivdeshpande** [comment]
> @sonalivdeshpande could you please describe the scope of this PR. It makes it easier to review it. Also, what are the use cases we concluded on that are driving MergeBranch implementation? I'm assuming you are trying to automate main/next here. I know we have discussed a couple of different approaches, I'm curious where we are at given the data we have so far.

I have updated the PR description

File: build-tools/packages/build-cli/src/api.ts [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+const GET_USER = "GET /repos/{owner}/{repo}/branches/{branch}/protection/restrictions/users";
+const DESCRIPTION = `
+        ## Main-next integrate PR
+        The aim of this pull request is to sync main and next branch. The expectation from the assignee is as follows:
+        > - Acknowledge the pull request by adding a comment -- "Actively working on it".
+        > - Resolve any merge conflicts between this branch and next (and push the resolution to this branch). Merge next into this branch if needed. **Do NOT rebase or squash this branch: its history must be preserved**.
+        > - Ensure CI is passing for this PR, fixing any issues. Please don't look into resolving **Real service e2e test** and **Stress test** failures as they are **non-required** CI failures.
+        For more information about how to resolve merge conflicts and CI failures, visit [this wiki page](https://github.com/microsoft/FluidFramework/wiki/Main-next-Automation).`;
+const TITLE = "Automate: Main Next Integrate";
+const log: Logger = defaultLogger;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/api.ts:27]: You probably want this to be passed in to the functions individually. Or otherwise make it settable externally.

When these functions are called from the cli code, they should log using the cli logger.

File: build-tools/packages/build-cli/src/api.ts [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+/**
+ *
+ * @param token - GitHub authentication token
+ * @returns Returns true if pull request exists
+ */
+export async function pullRequestExists(token: string): Promise<boolean> {
+    log.verbose("CHECKING IF PULL REQUEST EXISTS");
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/api.ts:35]: It's helpful to include what you're looking for in the log message. So if you're checking for  a PR with a title, include the title in the log message. It makes the logs more understandable on their own.

File: build-tools/packages/build-cli/src/api.ts [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    return response.data.some((d) => d.title === TITLE);
+}
+
+/**
+ *
+ * @param token - GitHub authentication token
+ * @param commit_sha - Commit id for which we need pull request information
+ */
+export async function pullRequestInfo(token: string, commit_sha: string): Promise<any> {
+    log.verbose("PULL REQUEST INFO");
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/api.ts:48]: Another place where you could include the sha or short sha to make the logs more useful.

File: build-tools/packages/build-cli/src/api.ts [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+    return prInfo;
+}
+
+/**
+ *
+ * @param token - GitHub authentication token
+ * @returns Lists the user who have push access to this branch
+ */
+export async function getUserAccess(token: string): Promise<any> {
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/api.ts:64]: Can this return something other than any?

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+    public async run(): Promise<void> {
+        const flags = this.processedFlags;
+
+        const context = await this.getContext();
+        const gitRepo = context.gitRepo;
+        const prExists: boolean = await api.pullRequestExists(flags.auth);
+
+        if (prExists) {
+            this.warn(`Open pull request exists`);
+            this.exit(-1);
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:55]:
```suggestion
            this.error(`Open pull request exists`);
```

And maybe add a link to the open PR?

File: build-tools/packages/build-cli/src/commands/merge/branches.ts [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+        }
+
+        // git pull main
+        // git pull next
+
+        const lastMergedCommit = await gitRepo.mergeBase(
+            `origin/${flags.source}`,
+            `origin/${flags.target}`,
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/commands/merge/branches.ts:63]: Assuming origin may not be safe here. I think you might be able to check the release code for an example of getting the upstream url and using that to verify the right branches/remote.

File: build-tools/packages/build-cli/src/lib/dangerfile.ts [hunk 21]
```diff
@@ -24,7 +24,7 @@ const adoConstants = {
 };
 
 const localReportPath = "./artifacts/bundleAnalysis";
-let logger: Logger;
+const logger: Logger = defaultLogger;
```
Comment (tylerbutler) [inline build-tools/packages/build-cli/src/lib/dangerfile.ts:27]: As mentioned above this should be settable. Otherwise callers can't use their own logger.

Comment (tylerbutler) [reply build-tools/packages/build-cli/src/lib/dangerfile.ts:27]: But it needs to be passed in from somewhere so the signatures of the functions may need to change, etc.

File: build-tools/packages/build-tools/src/index.ts [hunk 22]
```diff
@@ -20,7 +20,7 @@ export { exec, execNoError } from "./bumpVersion/utils";
 export { VersionBag } from "./bumpVersion/versionBag";
 export { FluidRepo } from "./common/fluidRepo";
 export { getResolvedFluidRoot } from "./common/fluidUtils";
-export { Logger, LoggingFunction } from "./common/logging";
```
Comment (tylerbutler) [inline build-tools/packages/build-tools/src/index.ts:23]: Don't export the default logger. It's only to be used within the legacy code.

**Review by tylerbutler — APPROVED** [review]
